### PR TITLE
Add finetuning scripts

### PR DESCRIPTION
**Pull request by nithinraok**
# What does this PR do ?

Add a finetuning script along with config file for finetuning existing nemo models. 

**Collection**: ASR

# Changelog 
- Added speech_to_text_finetune.py script
- Added speech_to_text_finetune.yaml config 
- Jenkins run
- Doc strings

TODO:
* Update tutorial

# Usage
Eg: without changing tokenizer
```bash
 python speech_to_text_finetune.py \
 init_from_pretrained_model=stt_en_fastconformer_hybrid_large_pc \
 model.tokenizer.update_tokenizer=false
```
or
finetune with a new tokenizer (architecture of base model doesn't change)
```bash
 python speech_to_text_finetune.py  \
 init_from_pretrained_model=stt_en_fastconformer_hybrid_large_pc \
 model.tokenizer.update_tokenizer=true \
 model.tokenizer.dir=<path_to_tokenizer_dir> \
 model.tokenizer.type=<bpe/char>
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: examples/asr/conf/fastconformer/fast-conformer_transducer_bpe.yaml [hunk 1]
```diff
@@ -254,7 +254,6 @@ trainer:
   precision: 32 # 16, 32, or bf16
   log_every_n_steps: 10  # Interval of logging.
   enable_progress_bar: True
-  resume_from_checkpoint: null # The path to a checkpoint file to continue the training, restores the whole state including the epoch, step, LR schedulers, apex, etc.
```
Comment (titu1994) [inline examples/asr/conf/fastconformer/fast-conformer_transducer_bpe.yaml:257]: Were these configs not updated in the PTL 2 update ?

File: examples/asr/conf/speech_to_text_finetune.yaml [hunk 2]
```diff
@@ -0,0 +1,115 @@
+name: "Speech_To_Text_Finetuning"
+
+# use `init_from_nemo_model` or `init_from_pretrained_model` to initialize the model
+# We do not currently support `init_from_ptl_ckpt` to create a single script for all types of models.
```
Comment (titu1994) [inline examples/asr/conf/speech_to_text_finetune.yaml:4]: We actually could support it in a roundabout way - of ckpt file, load the ptl ckpt with torch loss, read the config embedded in it, that will have target class path - we can resolve it and use that class to call load from checkpoint.

File: examples/asr/conf/speech_to_text_finetune.yaml [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  skip_nan_grad: false
+
+  train_ds:
+    manifest_filepath: ???
+    sample_rate: ${model.sample_rate}
+    batch_size: 16 # you may increase batch_size if your memory allows
+    shuffle: true
+    num_workers: 8
+    pin_memory: true
+    max_duration: 16.7 # it is set for LibriSpeech, you may need to update it for your dataset
```
Comment (titu1994) [inline examples/asr/conf/speech_to_text_finetune.yaml:22]: Let's keep 20 ?

File: examples/asr/conf/speech_to_text_finetune.yaml [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    num_workers: 8
+    pin_memory: true
+  
+  tokenizer:
+    update_tokenizer: false
+    dir: null  # path to directory which contains either tokenizer.model (bpe) or vocab.txt (for wpe)
+    type: bpe  # Can be either bpe (SentencePiece tokenizer) or wpe (WordPiece tokenizer)
+
+  spec_augment:
+    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
```
Comment (titu1994) [inline examples/asr/conf/speech_to_text_finetune.yaml:56]: Do we want a book like "use_specaug"

File: examples/asr/conf/speech_to_text_finetune.yaml [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    freq_masks: 2 # set to zero to disable it
+    time_masks: 10 # set to zero to disable it
+    freq_width: 27
+    time_width: 0.05
+
+  optim:
+    name: adamw
+    lr: 5e-3
```
Comment (titu1994) [inline examples/asr/conf/speech_to_text_finetune.yaml:64]: Too high ? It's 5e-4 or 0.001 usually for finetuning

File: examples/asr/conf/speech_to_text_finetune.yaml [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  exp_dir: null
+  name: ${name}
+  create_tensorboard_logger: true
+  create_checkpoint_callback: true
+  checkpoint_callback_params:
+    # in case of multiple validation sets, first one is used
+    monitor: "val_wer"
+    mode: "min"
+    save_top_k: 5
+    always_save_nemo: True # saves the checkpoints as nemo files instead of PTL checkpoints
```
Comment (titu1994) [inline examples/asr/conf/speech_to_text_finetune.yaml:108]: Hmm not instead of but along with. We have both Nemo and ckpt files

File: examples/asr/speech_to_text_finetune.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+
+@hydra_runner(config_path="conf", config_name="speech_to_text_finetune")
+def main(cfg):
+    logging.info(f'Hydra config: {OmegaConf.to_yaml(cfg)}')
+
+    trainer = pl.Trainer(**cfg.trainer)
+    exp_manager(trainer, cfg.get("exp_manager", None))
+
+    if hasattr(cfg, 'init_from_ptl_ckpt') and cfg.init_from_ptl_ckpt is not None:
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:72]: Let's move stuff to different functions to keep things clean

File: examples/asr/speech_to_text_finetune.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+        return asr_model
+
+    asr_model = get_base_model(cfg)
+    vocab_size = asr_model.tokenizer.vocab_size
+
+    # if new tokenizer is provided, use it
+    if hasattr(cfg.model.tokenizer, 'update_tokenizer') and cfg.model.tokenizer.update_tokenizer:
+        decoder = copy.deepcopy(asr_model.decoder)
+        joint_state = copy.deepcopy(asr_model.joint)
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:101]: First check if model has a joint or not. We could maybe make this a universal finetuning script then for CTC RNNT and hybrid models.

File: examples/asr/speech_to_text_finetune.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        elif pretrained_name is not None:
+            asr_model = ASRModel.from_pretrained(model_name=pretrained_name)
+
+        return asr_model
+
+    asr_model = get_base_model(cfg)
+    vocab_size = asr_model.tokenizer.vocab_size
+
+    # if new tokenizer is provided, use it
+    if hasattr(cfg.model.tokenizer, 'update_tokenizer') and cfg.model.tokenizer.update_tokenizer:
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:99]: Move Tokenizer stuff and such sections to their own functions

File: examples/asr/speech_to_text_finetune.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        logging.info("Using the tokenizer provided through config")
+        asr_model.change_vocabulary(
+            new_tokenizer_dir=cfg.model.tokenizer.dir, new_tokenizer_type=cfg.model.tokenizer.type
+        )
+        if asr_model.tokenizer.vocab_size != vocab_size:
+            logging.warning(
+                "The vocabulary size of the new tokenizer differs from that of the loaded model. As a result, finetuning will proceed with the new vocabulary, and the decoder will be reinitialized."
+            )
+        else:
+            asr_model.decoder = decoder
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:114]: Not like this. Update the state dict, not the full module

File: examples/asr/speech_to_text_finetune.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            asr_model = ASRModel.from_pretrained(model_name=pretrained_name)
+
+        return asr_model
+
+    asr_model = get_base_model(cfg)
+    vocab_size = asr_model.tokenizer.vocab_size
+
+    # if new tokenizer is provided, use it
+    if hasattr(cfg.model.tokenizer, 'update_tokenizer') and cfg.model.tokenizer.update_tokenizer:
+        decoder = copy.deepcopy(asr_model.decoder)
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:100]: Do NOT deep copy pytorch modules. Only call module.state_dict() and use load_state_dict() latef. 

Our modules are not safe to deep copy

File: examples/asr/speech_to_text_finetune.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        else:
+            asr_model.decoder = decoder
+            asr_model.joint = joint_state
+    else:
+        logging.info("Reusing the tokenizer from the loaded model.")
+
+    # Setup Data
+    cfg = model_utils.convert_model_config_to_dict_config(cfg)
+    asr_model.setup_training_data(cfg.model.train_ds)
+    asr_model.setup_validation_data(cfg.model.validation_ds)
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:122]: Setup_multiple_*

File: examples/asr/speech_to_text_finetune.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    if hasattr(cfg.model, 'test_ds') and cfg.model.test_ds.manifest_filepath is not None:
+        asr_model.setup_test_data(cfg.model.test_ds)
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:124]: ^

File: examples/asr/speech_to_text_finetune.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+                "The vocabulary size of the new tokenizer differs from that of the loaded model. As a result, finetuning will proceed with the new vocabulary, and the decoder will be reinitialized."
+            )
+        else:
+            asr_model.decoder = decoder
+            asr_model.joint = joint_state
+    else:
+        logging.info("Reusing the tokenizer from the loaded model.")
+
+    # Setup Data
+    cfg = model_utils.convert_model_config_to_dict_config(cfg)
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:120]: Here onward, these are one liners, probably no need to put into functions. Keep them as is

**Review by titu1994 — CHANGES_REQUESTED** [review]
This is really fantastic work ! I think with a bit of updating, we can make it a universal script that we can put under examples/asr right next to the eval and transcription scripts. 

Another thing, we need to add link to this in the asr ctc finetuming tutorial, plus add a section in the docs for finetuning and point to this.

File: examples/asr/conf/fastconformer/fast-conformer_transducer_bpe.yaml (hunk 1 above)
Comment (nithinraok) [reply examples/asr/conf/fastconformer/fast-conformer_transducer_bpe.yaml:257]: No, actually all other asr and even SV and SD configs need to be updated too.

File: examples/asr/conf/speech_to_text_finetune.yaml (hunk 2 above)
Comment (nithinraok) [reply examples/asr/conf/speech_to_text_finetune.yaml:4]: I thought about that too, but felt its not beautiful. May be I can put this in modelPT and use it here? So these functions can be used elsewhere.

File: examples/asr/conf/speech_to_text_finetune.yaml (hunk 4 above)
Comment (nithinraok) [reply examples/asr/conf/speech_to_text_finetune.yaml:56]: Ah no, since replacing values with 0, would turn off spec aug anyway.

File: examples/asr/conf/speech_to_text_finetune.yaml (hunk 5 above)
Comment (nithinraok) [reply examples/asr/conf/speech_to_text_finetune.yaml:64]: Ah good point, haven;t focused into details of trainer.

File: examples/asr/conf/speech_to_text_finetune.yaml (hunk 6 above)
Comment (nithinraok) [reply examples/asr/conf/speech_to_text_finetune.yaml:108]: copy pasted from FC configs, need to update there as well

File: examples/asr/speech_to_text_finetune.py (hunk 7 above)
Comment (nithinraok) [reply examples/asr/speech_to_text_finetune.py:72]: makes sense, to utils or modelPT?

Comment (titu1994) [reply examples/asr/speech_to_text_finetune.py:72]: No, functions inside this script is fine probably. If later it turns out to be generic and useful we can pull it out into utils.

File: examples/asr/conf/speech_to_text_finetune.yaml (hunk 4 above)
Comment (titu1994) [reply examples/asr/conf/speech_to_text_finetune.yaml:56]: Ah true.

File: examples/asr/conf/fastconformer/fast-conformer_transducer_bpe.yaml (hunk 1 above)
Comment (titu1994) [reply examples/asr/conf/fastconformer/fast-conformer_transducer_bpe.yaml:257]: This is odd - why were they skipped and how did tests pass then? @athitten

Comment (athitten) [reply examples/asr/conf/fastconformer/fast-conformer_transducer_bpe.yaml:257]: @titu1994 during the upgrade we just removed it for the yaml files that complained of this error. If you still see them in the some of the configs it's probably bcoz they were not run in the CI and weren't caught.

Comment (nithinraok) [reply examples/asr/conf/fastconformer/fast-conformer_transducer_bpe.yaml:257]: for tests, we use temp created configs, but It should be caught in tutorials

File: examples/asr/speech_to_text_finetune.py (hunk 8 above)
Comment (nithinraok) [reply examples/asr/speech_to_text_finetune.py:101]: Very good point, I missed this.

File: examples/asr/speech_to_text_finetune.py (hunk 12 above)
Comment (nithinraok) [reply examples/asr/speech_to_text_finetune.py:122]: moved to setup_dataloaders()

File: docs/source/asr/configs.rst [hunk 15]
```diff
@@ -1082,7 +1088,16 @@ Fine-tuning via a Pytorch Lightning checkpoint
         trainer.devices=-1 \
         trainer.accelerator='gpu' \
         trainer.max_epochs=50 \
-        +init_from_ptl_ckpt="<name of pytorch lightning checkpoint>"
+        +init_from_nemo_model="<path to .nemo model file>" # (or +init_from_pretrained_model, +init_from_ptl_ckpt )
```
Comment (titu1994) [inline docs/source/asr/configs.rst:1091]: Init from ptl ckpt is supported for this function, why is it being removed ?

File: examples/asr/speech_to_text_finetune.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+    elif nemo_model_path is not None:
+        asr_model = ASRModel.restore_from(restore_path=nemo_model_path)
+    elif pretrained_name is not None:
+        asr_model = ASRModel.from_pretrained(model_name=pretrained_name)
+
+    return asr_model
+
+
+def update_tokenizer(asr_model, tokenizer_dir, tokenizer_type):
+    vocab_size = asr_model.tokenizer.vocab_size
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:83]: Can we make no assumption about Char vs Subword models ? Lots of people seem to use char models

File: examples/asr/speech_to_text_finetune.py [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+        if joint_state is not None:
+            asr_model.joint.load_state_dict(joint_state)
+
+    return asr_model
+
+
+def setup_dataloaders(asr_model, cfg):
+    cfg = model_utils.convert_model_config_to_dict_config(cfg)
+    asr_model.setup_training_data(cfg.model.train_ds)
+    asr_model.setup_validation_data(cfg.model.validation_ds)
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:109]: Setup multiple validation data

File: examples/asr/speech_to_text_finetune.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+    if hasattr(cfg.model, 'test_ds') and cfg.model.test_ds.manifest_filepath is not None:
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:110]: Same as above

File: examples/asr/speech_to_text_finetune.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+from pytorch_lightning.utilities import rank_zero_only
+
+from nemo.collections.asr.models import ASRModel
+from nemo.core.config import hydra_runner
+from nemo.utils import logging, model_utils
+from nemo.utils.exp_manager import exp_manager
+
+
+@rank_zero_only
+def get_base_model(cfg):
```
Comment (titu1994) [inline examples/asr/speech_to_text_finetune.py:64]: Need docstrings everywhere

**Review by titu1994 — COMMENTED** [review]
Looks fantastic ! One thing is we need to support char models

File: docs/source/asr/configs.rst (hunk 15 above)
Comment (nithinraok) [reply docs/source/asr/configs.rst:1091]: I just updated the doc to characterize finetuning based on tokenizer updating or model architecture updating. 
Hence condensed all the finetuning docs to one subsection, else it looks lengthy.

File: examples/asr/speech_to_text_finetune.py (hunk 17 above)
Comment (nithinraok) [reply examples/asr/speech_to_text_finetune.py:109]: oh yeah

File: examples/asr/speech_to_text_finetune.py (hunk 18 above)
Comment (nithinraok) [reply examples/asr/speech_to_text_finetune.py:110]: Done

**Review by titu1994 — APPROVED** [review]
Looks great ! Thanks !
